auth/cache.go

```diff
@@ -0,0 +1,59 @@
+package auth
+
+import (
+	"math/rand"
+	"sync"
+)
+
+// Cache represents a random replacement cache.
+type Cache struct {
+	size  int                 // Maximum size where this cache can potentially grow upto.
+	keys  []*string           // A slice of keys for choosing a random key for eviction.
+	cache map[string]struct{} // Set of keys for fast lookup.
+	lock  sync.RWMutex        // Protects both cache and keys from concurrent access.
+}
+
+// Creates a new cache that can potentially grow upto the provided size.
+func NewCache(size int) *Cache {
+	return &Cache{
+		size:  size,
+		cache: make(map[string]struct{}),
```

Comment:
Preallocate size for the map and slice (to avoid allocs/copying when they need resizing on append).

```suggestion
		cache: make(map[string]struct{}, size),
		keys:  make([]*string, 0, size),
```

Reply:
This was considered initially but not included in the first place because even though we know that the maximum size of the cache is going to be up to 25000, it might be possible that this cache is not filled entirely when there are less than 25000 concurrent users, and I thought pre allocation of memory can impact the garbage collection.

Reply:
Fair point about the size being a cap instead of expected size. Makes sense, thanks.